### PR TITLE
feat(Tags): Tag component evolution

### PR DESCRIPTION
**Pull request by pladaria**
https://www.figma.com/file/wseConngNv2MYEXgCk5uI4/Tags-Specs?node-id=0%3A1

### CODE REVIEW

File: src/skins/blau.tsx [hunk 1]
```diff
@@ -167,6 +170,22 @@ export const getBlauSkin: GetSkin = () => {
...
             textAppBarSelected: palette.blauBlueSecondary60,
+
+            // TAGS
+            tagBackgroundSuccess: palette.blauGreen10,
+            tagBackgroundWarning: palette.blauYellow10,
+            tagBackgroundError: palette.blauRed10,
+            tagBackgroundPromo: palette.blauPurple10,
+            tagBackgroundActive: palette.blauBlueSecondary10,
+            tagBackgroundInactive: palette.grey2,
+            tagBackgroundInverse: palette.white,
```
Comment (pladaria) [inline src/skins/blau.tsx:181]: @yceballost this color (`tagBackgroundInverse`) is defined in the tags spec doesn't exist in the schema
https://github.com/Telefonica/mistica-design/blob/production/tokens/vivo-skin-schema.json

Comment (yceballost) [reply src/skins/blau.tsx:181]: my mistake, use `inverse`. Specs corrected!

File: src/tag.tsx [hunk 2]
```diff
@@ -2,35 +2,88 @@ import * as React from 'react';
...
+            style={{background: shouldUseInverseBackground ? colors.tagBackgroundInverse : backgroundColor}}
+        >
+            {Icon && (
+                <div style={{paddingRight: 4}}>
+                    <Icon color={textColor} size={pxToRem(16)} />
+                </div>
+            )}
             <ThemeVariant isInverse={false}>
-                <Text color={textColor} size={12} lineHeight={16} weight="regular" uppercase>
+                <Text color={textColor} size={14} lineHeight={20} weight="medium" truncate>
```
Comment (yceballost) [inline src/tag.tsx:86]: missing `text-transform: capitalize;` style for labels. Label always are capitalized

File: src/__stories__/data-card-story.tsx [hunk 3]
```diff
@@ -35,7 +35,7 @@ export default {
```
Comment (yceballost) [inline src/__stories__/data-card-story.tsx:35]: Data Card and Media card have new space between Tag and Headline. (+8px, 16px in total)

File: src/__stories__/tag-story.tsx [hunk 4]
```diff
@@ -1,42 +1,139 @@
...
+    IconAlarmLightRegular,
+    IconLikeRegular,
+    IconSkullRegular,
+    IconSmileySadRegular,
+    Inline,
+    Text1,
+} from '..';
+import {ThemeVariant} from '../theme-variant-context';
+
+type Args = {
```
Comment (yceballost) [inline src/__stories__/tag-story.tsx:18]: Is it possible to show list of args in playroom?
image

Comment (pladaria) [reply src/__stories__/tag-story.tsx:18]: oki

File: src/tag.tsx (hunk 2 above)
Comment (pladaria) [reply src/tag.tsx:86]: ok

**Review by yceballost — CHANGES_REQUESTED** [review]

File: src/skins/blau.tsx (hunk 1 above)
Comment (pladaria) [reply src/skins/blau.tsx]: done

File: src/__stories__/data-card-story.tsx (hunk 3 above)
Comment (pladaria) [reply src/__stories__/data-card-story.tsx:35]: done

File: src/skins/movistar.tsx [hunk 5]
```diff
@@ -64,72 +81,72 @@ export const getMovistarSkin: GetSkin = (variant) => {
...
             buttonDangerBackground: palette.pepper,
-            buttonDangerBackgroundDisabled: palette.pepperLight30,
-            buttonDangerBackgroundSelected: palette.pepperDark,
-            buttonDangerBackgroundHover: palette.pepperDark, // web only
+            buttonDangerBackgroundDisabled: palette.pepper20,
+            buttonDangerBackgroundSelected: palette.pepper55,
+            buttonDangerBackgroundHover: palette.pepper55, // web only
 
-            buttonLinkBackgroundSelected: palette.movistarBlueLight10,
+            buttonLinkBackgroundSelected: palette.movistarBlue20,
```
[the remark this answers is not in the document]
Comment (pladaria) [reply src/skins/movistar.tsx]: good catch, fixed

File: playroom/snippets.tsx [hunk 6]
```diff
@@ -623,6 +623,14 @@ const titlesSnippets: Array<Snippet> = [
     },
 ].map((snippet) => ({...snippet, group: 'Titles'}));
 
+const tagSnippets: Array<Snippet> = ['promo', 'active', 'inactive', 'success', 'warning', 'error'].map(
```
Comment (yceballost) [inline playroom/snippets.tsx:626]: You did Almendruco's trick and now you can't put the first letter in uppercase in snippets, right? 😬

Comment (pladaria) [reply playroom/snippets.tsx:626]: sure you can

Comment (pladaria) [reply playroom/snippets.tsx:628]: done

File: src/tag.tsx [hunk 7]
```diff
@@ -2,35 +2,95 @@ import * as React from 'react';
...
+type IconProps = {color?: string; size?: string | number};
+
+type TagType = 'promo' | 'active' | 'inactive' | 'success' | 'warning' | 'error';
+
 export type TagProps = {
+    // not using "TagType" and repeating the union to make these props playroom autocomplete friendly
+    type?: 'promo' | 'active' | 'inactive' | 'success' | 'warning' | 'error';
     children: string;
-    color: string;
+    Icon?: React.FC<IconProps>;
```
Comment (yceballost) [inline src/tag.tsx:33]: why `Icon` prop is capitalized?
image

Comment (pladaria) [reply src/tag.tsx:33]: Because it represents a component and components use PascalCase

You would use the Tag component as:
```js
<Tag Icon={IconPatata}>Patata</Tag>
```

instead of:
```js
<Tag icon={<IconPatata size="16" color="someColor" />}>Patata</Tag>
```

This way the Tag component is who decides how to render the icon (with a specific size and color)

Comment (yceballost) [reply src/tag.tsx:33]: could be more intutive maybe... `iconName` ?

`<Tag iconName={IconPatataRegular}>Patata</Tag>`

File: src/skins/o2-classic.tsx [hunk 8]
```diff
@@ -96,14 +121,14 @@ export const getO2ClassicSkin: GetSkin = () => {
             textButtonPrimaryInverseSelected: palette.o2SkyBlue,
             textButtonSecondary: palette.o2SkyBlue,
             textButtonSecondaryDisabled: palette.o2SkyBlueLight50,
```
[the remark this answers is not in the document]
Comment (AnaMontes11) [reply src/skins/o2-classic.tsx:123]: Substitute all o2SkyBlueLight50 colours for o2SkyBlueLight45

File: src/skins/o2-classic.tsx [hunk 9]
```diff
@@ -3,41 +3,66 @@ import {O2_CLASSIC_SKIN} from './constants';
...
+    o2Blue55: '#04264E',
+    o2Blue45: '#6C8BAF',
+    o2Blue10: '#E6EAEE',
+
     o2SkyBlue: '#0090D0',
-    o2SkyBlueDark: '#057DB2',
+    o2SkyBlue55: '#057DB2',
+
     o2SkyBlueLight: '#65B4E4',
     o2SkyBlueLight50: '#80C7E7',
```
[the remark this answers is not in the document]
Comment (AnaMontes11) [reply src/skins/o2-classic.tsx:19]: Substitute all o2SkyBlueLight50 colours for o2SkyBlueLight45

File: src/tag.tsx [hunk 10]
```diff
@@ -2,35 +2,95 @@ import * as React from 'react';
...
-    // Hardcode black text in darkmode because there isn't a black text color constant that we can use in dark mode
-    const blackText = isDarkMode ? '#313235' : colors.textPrimary;
-    const textColor = color === colors.inverse ? blackText : colors.textPrimaryInverse;
+    const isInverse = useIsInverseVariant();
+
+    /**
+     * Legacy implementation
+     *
+     * @deprecated to be removed in the next major version
+     */
```
Comment (atabel) [inline src/tag.tsx:48]: move this comment down

File: src/tag.tsx [hunk 11]
```diff
@@ -2,35 +2,95 @@ import * as React from 'react';
...
+    const shouldUseInverseBackground = isInverse && !isDarkMode;
+
     return (
-        <span className={classes.tag}>
+        <span
+            className={classes.tag}
+            style={{background: shouldUseInverseBackground ? colors.inverse : backgroundColor}}
+        >
+            {Icon && (
+                <div style={{paddingRight: 4}}>
```
Comment (atabel) [inline src/tag.tsx:88]: Box?

File: src/tag.tsx [hunk 12]
```diff
@@ -2,35 +2,95 @@ import * as React from 'react';
...
+                    <Icon color={textColor} size={pxToRem(16)} />
+                </div>
+            )}
             <ThemeVariant isInverse={false}>
-                <Text color={textColor} size={12} lineHeight={16} weight="regular" uppercase>
+                <Text color={textColor} size={14} lineHeight={20} weight="medium" truncate>
```
Comment (atabel) [inline src/tag.tsx:93]: is this what we want? can't we use a `Text2`? we don't want `Tags` to have bigger font size in desktop?

File: src/__stories__/media-card-story.tsx [hunk 13]
```diff
@@ -21,22 +20,17 @@ export default {
...
-        warning: colors.warning,
-        error: colors.error,
+    const tagTypes = {
+        promo: 'promo',
+        active: 'active',
+        inactive: 'inactive',
+        success: 'success',
+        warning: 'warning',
+        error: 'error',
     };
```
Comment (atabel) [inline src/__stories__/media-card-story.tsx:30]: you can use an array now

**Review by atabel — COMMENTED** [review]
please capitalize Tags in cards stories:
image

File: src/skins/movistar.tsx [hunk 14]
```diff
@@ -64,72 +81,72 @@ export const getMovistarSkin: GetSkin = (variant) => {
...
             buttonLinkBackgroundSelectedInverse: applyAlpha(palette.white, 0.2),
 
             buttonPrimaryBackground: palette.movistarGreen,
-            buttonPrimaryBackgroundDisabled: palette.movistarGreenLight50,
-            buttonPrimaryBackgroundDisabledInverse: palette.movistarBlueLight50,
+            buttonPrimaryBackgroundDisabled: palette.movistarGreen30,
+            buttonPrimaryBackgroundDisabledInverse: palette.movistarBlue10,
```
[the remark this answers is not in the document]
Comment (pladaria) [reply src/skins/movistar.tsx]: I think you're commenting on an outdated file

File: src/skins/o2-classic.tsx [hunk 15]
```diff
@@ -3,41 +3,66 @@ import {O2_CLASSIC_SKIN} from './constants';
...
+    o2SkyBlueLight45: '#80C7E7',
     o2SkyBlueLight30: '#D0E8F6',
```
[the remark this answers is not in the document]
Comment (pladaria) [reply src/skins/o2-classic.tsx:21]: done, thanks!

File: src/skins/o2-classic.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (pladaria) [reply src/skins/o2-classic.tsx:19]: done, thanks!

File: src/skins/movistar.tsx [hunk 16]
```diff
@@ -3,39 +3,56 @@ import {MOVISTAR_SKIN} from './constants';
...
     movistarProminentBlueDark: '#081F2D',
     movistarProminentBlueLight20: '#CED3D7',
     movistarProminentBlueLight50: '#85939C',
     movistarProminentBlueLight70: '#546874',
+
     // specific for dark mode:
     darkModeBlack: '#191919',
     darkModeGrey: '#242424',
-    darkModeMovistarBlueDark: '#4C7389',
+    darkModemovistarBlue55: '#4C7389',
```
[the remark this answers is not in the document]
Comment (pladaria) [reply src/skins/movistar.tsx]: done!

File: src/tag.tsx (hunk 12 above)
Comment (pladaria) [reply src/tag.tsx:93]: Yes, this is documented in the spec:
image

File: src/__stories__/media-card-story.tsx (hunk 13 above)
Comment (pladaria) [reply src/__stories__/media-card-story.tsx:30]: done

File: src/tag.tsx (hunk 10 above)
Comment (pladaria) [reply src/tag.tsx:48]: done

File: src/tag.tsx (hunk 7 above)
Comment (pladaria) [reply src/tag.tsx:34]: `iconName` is less intuitive, I would expect that prop to be a string.

A more obvious name would be `IconComponent` but I prefer the short version

File: src/button.tsx [hunk 17]
```diff
@@ -383,6 +383,12 @@ const Button: React.FC<ButtonProps & {classes: ReturnType<typeof usePrimaryButto
         role: 'button',
     };
```
Comment (pladaria) [inline src/button.tsx:385]: unrelated changes to make button more resilient, requested by @Winde

File: src/tag.tsx (hunk 12 above)
Comment (atabel) [reply src/tag.tsx:93]: a comment in the code would be nice

File: src/button.tsx (hunk 17 above)
Comment (atabel) [reply src/button.tsx:385]: we could consider using some `invariant` [library](https://www.npmjs.com/package/tiny-invariant) for this kind of checks. We have these checks in other components too. TS/Flow understand `invariant` calls and refine the types too.

**Review by atabel — APPROVED** [review]

**Review by yceballost — APPROVED** [review]

**Review by AnaMontes11 — APPROVED** [review]

**Review by atabel — APPROVED** [review]
